### ISSUE TITLE
Bug: sam build -u stuck on mounting

### ISSUE THREAD

**Issue by BFlores16** [aws/aws-sam-cli#7387]
### Description:

I have a Gitlab pipeline that builds and deploys my SAM application. My application contains about 30 lambdas with mostly python and some node. I have never had an issue when I run `sam build -u` locally. But when running the command in my pipeline, the pipeline hangs on the last function and gets stuck on "Mounting /builds/<rest_of_path>/ /tmp/samcli/source:ro,delegated, inside runtime container"

In order to resolve this, I have to delete all artifacts in my repo and then clear runner caches. Then the pipeline will work once with sam build, and then get stuck again on subsequent runs. I've tried modifying my gitlab-ci.yml in many ways with no success.

Here is my gitlab-ci.yml
```yaml
variables:
  SAM_TEMPLATE: Lambdas/template.yaml
  DOCKER_DRIVER: overlay2
  DOCKER_TLS_CERTDIR: "/certs"

services:
  - docker:23.0.6-dind

# Should always specify a specific version of the image. If using a tag like docker:stable,
# there will be no control over which version is used. Unpredictable behavior can result.
image: docker:23.0.6

before_script:
  - apk add --update python3 py-pip python3-dev build-base libffi-dev
  - pip install --upgrade pip
  - pip install awscli aws-sam-cli
  
stages:
  - preview
  - deploy 

preview:
  stage: preview
  script:
    - chmod 755 aws-variables.sh
    - ./aws-variables.sh
    - export AWS_DEFAULT_REGION=$AWS_REGION
    - cd Lambdas
    - sam build -u
    - sam deploy --region $AWS_REGION --no-execute-changeset --no-fail-on-empty-changeset
    - cd ..
    - changeset_id=$(aws cloudformation describe-change-set --stack-name Lambdas --change-set-name $(aws cloudformation list-change-sets --stack-name Lambdas --query "sort_by(Summaries, &CreationTime)[-1].ChangeSetName" --output text) --query "ChangeSetId" --output text)
    - echo $changeset_id > changeset.txt
  artifacts:
    paths:
      - changeset.txt

deploy-prod:
  stage: deploy
  script:
    - chmod 755 aws-variables.sh
    - ./aws-variables.sh
    - changeset_id=$(cat changeset.txt)
    - export AWS_DEFAULT_REGION=$AWS_REGION
    - aws cloudformation execute-change-set --change-set-name $changeset_id
  only:
    - main
    - develop
  when: manual
  environment:
    name: production

```

### Steps to reproduce:

1. Change code and push
2. Create merge request
3. Pipeline runs
4. Pipeline gets stuck on `sam build -u`

### Observed result:

Pipeline gets stuck on `sam build -u`

### Expected result:

Build should succeed and `sam deploy` should proceed

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: MacOS 13.1
5. `sam --version`: 1.123.0
6. AWS region: us-west-1

```
# Paste the output of `sam --info` here
```
{
  "version": "1.123.0",
  "system": {
    "python": "3.11.8",
    "os": "Linux-5.15.154+-x86_64-with"
  },
  "additional_dependencies": {
    "docker_engine": "23.0.6",
    "aws_cdk": "Not available",
    "terraform": "Not available"
  },
  "available_beta_feature_env_vars": [
    "SAM_CLI_BETA_FEATURES",
    "SAM_CLI_BETA_BUILD_PERFORMANCE",
    "SAM_CLI_BETA_TERRAFORM_SUPPORT",
    "SAM_CLI_BETA_RUST_CARGO_LAMBDA"
  ]
}

`Add --debug flag to command you are running`

**Comment by BFlores16** [comment]
> Hi @BFlores16 , thank you for reporting the issue.
> 
> It looks like to me that docker was running out of memory on the host that run `sam build -u`. Could you please provide us the infrastructure detail of the host that run the job? (CPU, RAM, Disk space, ...).
> 
> You can also try splitting into smaller batches of lambdas and see whether it is still failing.

I am using the GitLab shared runners, which should have the following configs:
- 1 vCPU
- 4 GB of RAM
- 10 GB of disk space
- Timeout 1 hour

Here's some info I output in my pipeline. 
```
$ echo "===== Memory Info ====="
===== Memory Info =====
$ free -h
               total        used        free      shared  buff/cache   available
Mem:           7.8Gi       569Mi       5.8Gi       1.0Mi       1.7Gi       7.2Gi
Swap:          2.0Gi          0B       2.0Gi
$ echo "===== Disk Space Info ====="
===== Disk Space Info =====
$ df -h
Filesystem                Size      Used Available Use% Mounted on
overlay                  25.4G      7.7G     17.6G  31% /
tmpfs                    64.0M         0     64.0M   0% /dev
shm                      64.0M         0     64.0M   0% /dev/shm
/dev/sda1                25.4G      7.7G     17.6G  31% /builds
/dev/sda1                25.4G      7.7G     17.6G  31% /certs/client
/dev/sda1                25.4G      7.7G     17.6G  31% /etc/resolv.conf
/dev/sda1                25.4G      7.7G     17.6G  31% /etc/hostname
/dev/sda1                25.4G      7.7G     17.6G  31% /etc/hosts
/dev/sda1                25.4G      7.7G     17.6G  31% /var/lib/docker
tmpfs                     3.9G         0      3.9G   0% /sys/devices/virtual/dmi/id
$ echo "===== Docker Info ====="
===== Docker Info =====
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/local/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.22.0
    Path:     /usr/local/libexec/docker/cli-plugins/docker-compose
Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 23.0.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.154+
 Operating System: Alpine Linux v3.18 (containerized)
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 7.768GiB
 Name: 08ef878f2db3
 ID: a2b3703f-1391-4b0a-8cc4-4d59185b43eb
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

**Comment by BFlores16** [comment]
The memory was captured during the pipeline run but prior to the sam build command. I don't think there is a way for me to evaluate the memory usage during the sam build command.

Sam build without the -u does not work for me because many of my lambdas are being containerized due to requirements.txt files. I may be ignorant on how best to package my dependencies and deploy them in the pipeline so if you have any suggestions I would welcome them please. Here's an example error without -u flag

Build Failed
Error: PythonPipBuilder:Validation - Binary validation failed for python, searched for python in following locations  : ['/usr/bin/python', '/usr/bin/python3'] which did not satisfy constraints for runtime: python3.9. Do you have python for runtime: python3.9 on your PATH?
Cleaning up project directory and file based variables
00:00
ERROR: Job failed: exit code 1

**Comment by dkphm** [comment]
Hi @BFlores16 , thank you for the prompt response. I would like to dive deeper into the logs to see what could went wrong, could you please sent us the full output of `sam build -u --debug` running on the pipeline?

Thanks!

**Comment by BFlores16** [comment]
Here is the log from a previous run with the --debug flag on

[sam-build-log.txt](https://github.com/user-attachments/files/16700949/sam-build-log.txt)

**Comment by dkphm** [comment]
Thanks @BFlores16 for the logs. 

Looks like I was able to reproduce the issue, this is likely because one of the container didn't return to the main thread after finishing the build. We are working on the fix and hopefully it could be available in the next release.

In the meantime, if you could install Python (with the same version as defined in `template.yaml`) and run `sam build` without `-u`, I believe it will help you unblock the pipeline.

**Comment by BFlores16** [comment]
I ended up changing all my python lambdas to a consistent version and installed python and node on my container. I also removed the -u flag. These were the added benefits:
- Consistently successful builds
- Faster build time
- The dependencies are cached so now my pipeline recognizes only changes to the lambdas. Previously it was rebuilding each lambda that had dependencies.

I may check one day to see if the team ever actually fixes the bug reported but I doubt I would ever switch back to using the flag as it doesn't seem to give me any benefits.

Here is how I modified my pipeline file:
```yaml
variables:
  SAM_TEMPLATE: Lambdas/template.yaml
  DOCKER_DRIVER: overlay2
  DOCKER_TLS_CERTDIR: "/certs"

services:
  - docker:23.0.6-dind

image: docker:23.0.6

before_script:
  - apk add --update python3 py3-pip python3-dev build-base libffi-dev util-linux procps
  - apk add nodejs npm
  - if ! python3 --version | grep -q "3.11"; then apk add --repository=http://dl-cdn.alpinelinux.org/alpine/edge/community python3=3.11*; fi
  - ln -sf /usr/bin/python3 /usr/local/bin/python
  - ln -sf /usr/bin/node /usr/local/bin/node
  - pip install --upgrade pip
  - pip install awscli aws-sam-cli
  
stages:
  - preview
  - deploy 

preview:
  stage: preview
  timeout: 20m
  script:
    - chmod 755 aws-variables.sh
    - ./aws-variables.sh
    - export AWS_DEFAULT_REGION=$AWS_REGION
    - cd Lambdas
    - sam build
    - sam deploy --region $AWS_REGION --no-execute-changeset --no-fail-on-empty-changeset
    - cd ..
    - changeset_id=$(aws cloudformation describe-change-set --stack-name Lambdas --change-set-name $(aws cloudformation list-change-sets --stack-name Lambdas --query "sort_by(Summaries, &CreationTime)[-1].ChangeSetName" --output text) --query "ChangeSetId" --output text)
    - echo $changeset_id > changeset.txt
  artifacts:
    paths:
      - changeset.txt

deploy-prod:
  stage: deploy
  script:
    - chmod 755 aws-variables.sh
    - ./aws-variables.sh
    - changeset_id=$(cat changeset.txt)
    - export AWS_DEFAULT_REGION=$AWS_REGION
    - aws cloudformation execute-change-set --change-set-name $changeset_id
  only:
    - main
    - develop
  when: manual
  environment:
    name: production
```

I would still be interested to see if this bug is ever fixed as it makes it easier to package dependencies as I won't have to install and maintain specific versions

**Comment by dkphm** [comment]
Hi @BFlores16 , thank you for the feedback.

Yes building without `-u` would be much faster and easier, but sometimes it is troublesome to maintain the environments and dependencies in all the hosts.

The issue has been added to our backlog and we will try to find a solution for this.

**Comment by roger-zhangg** [comment]
Hi, we're revisiting this issue during triage. It's been ~22 months since the last activity and several SAM CLI releases have shipped since then with container/Docker improvements.

Is anyone still experiencing `sam build -u` getting stuck on mounting with a recent SAM CLI version? If so, could you share:

1. Your SAM CLI version (`sam --version`)
2. Your Docker version (`docker --version`)
3. Your OS and architecture

If no one can still reproduce this, we'll close the issue in 14 days. Thanks!

**Comment by roger-zhangg** [comment]
Closing this as stale — there's been no response to the triage follow-up above in ~29 days, and the report is now well over two years old with many SAM CLI releases since, including container and Docker mount improvements.

If you're still hitting `sam build -u` getting stuck on mounting, please comment and **feel free to re-open** — no need to file a new issue. The details from the follow-up are what we'd need to act on it:

1. SAM CLI version (`sam --version`)
2. Docker version (`docker --version`)
3. Your OS and architecture

Thanks for the original report.
